api/src/main/java/org/apache/iceberg/io/ClosingIterator.java

```diff
@@ -25,33 +25,30 @@
 
 public class ClosingIterator<T> implements Iterator<T> {
   private final CloseableIterator<T> iterator;
-  private boolean shouldClose = false;
 
   public ClosingIterator(CloseableIterator<T> iterator) {
     this.iterator = iterator;
-
   }
 
   @Override
   public boolean hasNext() {
     boolean hasNext = iterator.hasNext();
-    this.shouldClose = !hasNext;
+    if (!hasNext) {
+      close();
```

Comment:
I think you mean in the read pattern (for the original `ClosingIterator`)

```java
   while (hasNext()) {
      T val = next();
    }
```

When the hasNext() return false, then we won't call the `next()` to close the internal `CloseableIterator`.   Yes, I think you are correct,  but I think we will need to guarantee that we will only close the `CloseableIterator` once even if we called the `ClosingIterator#hasNext` many times, otherwise once we've called the first `hasNext`, all the following `hasNext` check will be thrown an exception, that's not the expected behavior.

Reply:
`Closeable::close` is required to be idempotent - if calling close on a `Closeable` resource has any side-effects, we should fix that at the source. This would also mean that calling close() multiple times should not throw any exceptions.

Reply:
> we should fix that at the source

We can not assume that all the source could guarantee it only close one times, even if all the resources introduced in apache iceberg project don't guarantee it.  The correct approach is keeping it close once in this `ClosingIterator`. I think introducing a boolean `closed` is enough ?

Reply:
> We can not assume that all the source could guarantee it only close one times, even if all the resources introduced in apache iceberg project don't guarantee it.

I agree

Reply:
Done, and added a test.